remote-api/src/main/java/org/alfresco/rest/api/impl/DownloadsImpl.java

```diff
@@ -172,4 +198,99 @@ private Download getStatus(NodeRef downloadNodeRef)
         return downloadInfo;
     }
 
+    /**
+     * Checks the supplied nodes for any content that is archived.
+     * Any folders will be expanded and their children checked.
+     * A limit can be applied to prevent large sized requests preventing the asynchronous call to start.
+     * 
+     * @param nodeRefs
+     * @param checkLimit The maximum number of nodes to check, set to -1 for no limit
+     * @see #checkArchiveStatus(NodeRef[], List)
+     */
+    @Experimental
+    protected void checkArchiveStatus(NodeRef[] nodeRefs, int checkLimit) 
+    {
+        if (canCheckArchived())
+        {
+            checkArchiveStatus(nodeRefs, checkLimit, null);
+        }
+    }
+
+    /**
+     * Checks the supplied nodes for any content that is archived.
+     * Any folders will be expanded and their children checked.
+     * A limit can be applied to prevent large sized requests preventing the asynchronous call to start.
+     * The cache is used to prevent duplication of checks, as it is possible to provide a folder and its contents as 
+     * separate nodes in the download request.
+     * 
+     * @param nodeRefs
+     * @param checkLimit The maximum number of nodes to check, set to -1 for no limit
+     * @param cache Tracks nodes that we have already checked, if null an empty cache will be created
+     */
+    @Experimental
+    private void checkArchiveStatus(NodeRef[] nodeRefs, int checkLimit, List<NodeRef> cache)
+    {
+        // Create the cache for recursive calls.
+        if (cache == null) 
+        {
+            cache = new ArrayList<NodeRef>();
+        }
+
+        var folders = new ArrayList<NodeRef>();
+        for (NodeRef nodeRef : nodeRefs) 
+        {
+            // We hit the number of nodes we want to check.
+            if (cache.size() == checkLimit) 
+            {
+                logger.info(
+                        String.format(
+                                "Maximum check of %d reached for archived content. No more checks will be performed and download will still be created.",
+                                checkLimit));
+                return;
+            }
+            // Already checked this node, we can skip.

```

Comment:
How likely (when) will a node appear more than once (given we're only walking primary child assocs) ? 

Could we use two HashSet instead (eg. folderRefs and nonFolderRefs) and then only track a total count ?

Reply:
It's unlikely but possible. Image the following, simple, file heirarchy: folder1/folder2/content. The following download request is legal: 
{  
    "nodeIds": [    
        "node for folder1",
        "node for folder2",
        "node for content
    ]
}
You can see that we could attempt to look at the child refs of folder2 and check content multiple times.
I have attempted to replicate this complexity in the unit tests provided.

Reply:
Ah ok - thanks.